dataflow/src/ops/base.rs

```diff
@@ -129,6 +144,32 @@ fn key_of<'a>(key_cols: &'a [usize], r: &'a Record) -> impl Iterator<Item = &'a
         .map(move |(i, col)| key_val(i, *col, r))
 }
 
+// Replaces None values with the next available auto increment value, mutating
+// auto_increment_values if there are any changes.
+fn replace_with_auto_increment(
+    auto_increment_values: &mut HashMap<usize, DataType>,
+    row: &mut Vec<DataType>,
+    shard: Option<usize>,
+) {
+    for (index, value) in auto_increment_values.iter_mut() {
+        // When we're given a None value, replace it with the incremented version
+        // of the last auto increment value for that column (or the initial increment value):
+        *value = match (&row[*index], &value) {
+            (&DataType::None, &&mut DataType::None) => {
+                DataType::ID(shard.unwrap_or(0) as u32, INITIAL_AUTO_INCREMENT)
+            }
+            (&DataType::None, &&mut DataType::ID(s, i)) => DataType::ID(s, i + 1),
+            // Other values should override existing auto increment values, so that
+            // the auto incrementer continues from there the next time:
```

Comment:
Why? This seems like it could come back to bite us.. We should at least assert that the new `i` is higher than what we've stored..

Reply:
Was just to match MySQL behavior. Added an assertion, but would be happy to remove it altogether in favor of always requiring `DataType::None` as well.